### PR TITLE
Allow to specify units

### PR DESCRIPTION
**Pull request by hegner**
BEGINRELEASENOTES
- Allow to specify units as part of the datamodel definition 

ENDRELEASENOTES

This WIP is to allow the explicit inclusion of units into the datamodel definition. 

- [x] Extend the regular expressions/parsing to allow a unit description
- [x] Documentation

### CODE REVIEW

**Comment by hegner** [comment]
@tmadlener - can you let me know whether you'd agree with this? I needed some break from genreflex work and did this long-standing item on my wish list.

File: python/podio/test_MemberParser.py [hunk 1]
```diff
@@ -243,6 +243,13 @@ def test_parse_valid_no_description(self):
     self.assertEqual(parsed.description, 'descriptions are not ignored even though they are not required')
     self.assertTrue(not parsed.is_builtin)
 
+  def test_parse_unit(self):
+    """Test that units are properly parsed"""
+    parser = MemberParser()
+
+    parsed = parser.parse('unsigned long long var [GeV] // description')
+    self.assertEqual(parsed.unit, 'GeV')
+
```
Comment (tmadlener) [inline python/podio/test_MemberParser.py:252]: Could we add another test case here with an additional default value?

**Review by tmadlener — COMMENTED** [review]
I think this is a good idea. It is not yet clear to me if/how this will affect the generated code. That might have quite an impact on the size of this feature.

Would it mainly be an addition in the docstring, or would it actually enter in the generated c++ code. If the latter, which base units do we use, or do we leave that choice to the users (and if so, how)? Would we try to make this rather strict and introduce a proper units library to do all the conversions for us (there is a propoal for c++29 to have this in the standard library)?

File: python/podio/test_MemberParser.py (hunk 1 above)
Comment (hegner) [reply python/podio/test_MemberParser.py:252]: what would be a default value?

**Comment by hegner** [comment]
For the moment it is for documentation purposes. And I want to have a detection in the schema evolution in case it changes. I do not think we should use a units library unless being asked for explicitly.

File: python/podio/test_MemberParser.py (hunk 1 above)
Comment (tmadlener) [reply python/podio/test_MemberParser.py:252]: Yes, exactly. Having a default value and a unit. Just to cover all possibilities.

**Comment by tmadlener** [comment]
OK. That sounds reasonable to achieve then. Then the main comment/question I have is whether we also need this for the components?

**Comment by hegner** [comment]
Obviously :-)

**Comment by tmadlener** [comment]
Does this already work for components as well?

**Comment by hegner** [comment]
Yes it does. The test case covers `NotSoSimpleStruct` (component) and `ExampleHit` (datatype).

**Comment by tmadlener** [comment]
Ah right, sorry I missed that because it also has a description. Then more specifically does this also work with components that do not have a description, e.g. the `SimpleStruct`?

**Comment by hegner** [comment]
Sure. The description is default initialized as empty string (`""`), thus there is no special case to handle

**Comment by tmadlener** [comment]
I was wondering more about the yaml parsing side here. IIRC I special cased the components there to not require a description, and I am not sure if this already covers that special case

**Comment by hegner** [comment]
The combined regex would give a `None` for that part and we are back to default handling later. That's the nice thing of having unified the syntax of components and datatypes.

**Comment by tmadlener** [comment]
OK. IIUC, everything should work from the parsing side and nothing breaks if I add units, e.g. here:
https://github.com/AIDASoft/podio/blob/dbf9425aa6d2b956c4c6fa49d45abe40cbe9e5af/tests/datalayout.yaml#L13-L17

However, I think in the current implementation these units would also not end up in a docstring, because they are never passed to the `MemberVariable` in this case, right? (The `bare_<xyz>` regexes do not have the `unit_str` yet, and so the unit is also not passed to the `MemberVariable` constructor in the end).

**Comment by hegner** [comment]
@tmadlener - I addressed your comment with a fix. Indeed units were not handled correctly there.

**Comment by tmadlener** [comment]
It looks like the `clang12` based workflows are failing due to not having root in their environment. I suspect that these are transient issues in the LCG nightlies and they should be safe to ignore. The `key4hep` (release) based one is failing due to an unsuitable ROOT version, so failure there is also expected.

**Comment by andresailer** [comment]
The clang12 nightlies no longer exist

**Review by tmadlener — APPROVED** [review]

**Comment by tmadlener** [comment]
@hegner could you rebase this onto the latest master (or merge that in), so that we pick up a completely working CI again. After that finishes successfully, I would merge this.

**Comment by hegner** [comment]
@tmadlener - seems all green now :-)

**Review by tmadlener — APPROVED** [review]

**Pull request merged**